### ISSUE TITLE
Constant fold of tensor ops

### ISSUE THREAD

**Issue by okkwon** [iree-org/iree#9750]
### Request description

We may further fold constant by calculating tensor ops. Here is an example from PoseNet in the IREE benchmark suite.

<img width="880" alt="image" src="https://user-images.githubusercontent.com/4314836/178081719-fac771d2-7ed9-4af3-ba77-d8352e266bd8.png">

dispatch_71 is purely dependent on the constant inputs and can be folded into the <2xi32> constant value.

The implementation will be costly because we need to implement the arith operations as a library function call. (I am not aware of the feature in other dialects such as tosa or mhlo. If they provide the feature, we may fold the ops at that level.)

After folding the constant, we may further fuse dispatch 70 and 72.

### What component(s) does this issue relate to?

Compiler

**Comment by okkwon** [comment]
The pattern is repeating in the graph. [dispatch.pdf](https://github.com/iree-org/iree/files/9075456/dispatch.pdf)

**Comment by benvanik** [comment]
This should already be getting folded - we have a whole thing that does this constant evaluation and hosting - it's likely just getting missed.

If the hoist into globals pass is running and that isn't getting hoisted then that's likely the issue: https://github.com/iree-org/iree/blob/06bd50a6c69dde29a1fdb1525f21166b897f9d4e/compiler/src/iree/compiler/Dialect/Flow/Transforms/Passes.cpp#L139-L145

**Comment by okkwon** [comment]
Thanks Ben. I will take a look.

**Comment by stellaraccident** [comment]
We don't have those flags on by default as they need more testing and tuning and was waiting for good opportunities: https://github.com/iree-org/iree/blob/main/docs/website/docs/reference/optimization-options.md

You need both `--iree-opt-const-expr-hoisting` and `--iree-opt-const-eval`.

**Comment by benvanik** [comment]
Good point stella! iree-benchmark-module doesn't measure init time and so just doing the const-expr-hoisting and running the benchmarks should show whether it's worth it to potentially const eval something. So start with making sure the hoisting is working and then the tradeoff of binary size vs runtime init cost is something orthogonal.
